tck/src/main/java/org/eclipse/microprofile/lra/tck/participant/api/RecoveryResource.java

```diff
@@ -0,0 +1,187 @@
+import java.net.URI;
+import java.time.temporal.ChronoUnit;
+import java.util.logging.Logger;
+
+import static org.eclipse.microprofile.lra.annotation.ws.rs.LRA.LRA_HTTP_CONTEXT_HEADER;
+
+@Path(RecoveryResource.RECOVERY_RESOURCE_PATH)
+@ApplicationScoped
+public class RecoveryResource {
+
+    public static final String RECOVERY_RESOURCE_PATH = "recovery-resource";
+    public static final String PHASE_1 = "phase-1";
+    public static final String PHASE_2 = "phase-2";
+    public static final String LRA_HEADER = "LRA";
+    public static final String TRIGGER_RECOVERY = "triggerRecovery";
+    public static final String TIMEOUT_HEADER = "timeout-millis";
+    private static final String REQUIRED_PATH = "required";
+    private static final String REQUIRED_TIMEOUT_PATH = "required-timeout";
+
+    private static final Logger LOGGER = Logger.getLogger(RecoveryResource.class.getName());
+    
+    @Inject
+    LRAMetricService lraMetricService;
+
+    @Inject
+    private LraRecoveryService lraRecoveryService;
+
+    @Inject
+    LRATestService lraTestService;
+
+    @Inject
+    LraTckConfigBean lraTckConfigBean;
+
+    @PostConstruct
+    public void postConstruct() {
+        lraTestService.init();
+    }
+
+    @PreDestroy
+    public void preDestroy() {
+        lraTestService.clean();
+    }
+
+    /**
+     * Starts a new LRA and enlists an instance of this class with it as a participant
+     *
+     * @param timeout whether or not the LRA should time out after 500 milliseconds
+     * @return response containing the LRA id of the newly started LRA
+     */
+    @GET
+    @Path(PHASE_1)
+    public Response phase1(@QueryParam("timeout") @DefaultValue("false") boolean timeout) {
+        // start a new LRA and join it with this resource
+        URI lra;
+        Response response;
+        
+        response = lraTestService.getTCKSuiteTarget().path(RecoveryResource.RECOVERY_RESOURCE_PATH)
+                .path(timeout ? RecoveryResource.REQUIRED_TIMEOUT_PATH : RecoveryResource.REQUIRED_PATH)
+                .request().put(Entity.text(""));
+
+        lra = URI.create(response.readEntity(String.class));
+
+        Response.ResponseBuilder responseBuilder = Response.ok(lra);
+        if (timeout) {
+            responseBuilder.header(TIMEOUT_HEADER, lraTckConfigBean.getLongConsistencyDelay());
+        }
+        
+        return responseBuilder.build();
+    }
+
+    /**
+     * Cancels the supplied LRA and verifies that all required actions have been performed
+     * 
+     * @param lraId lra id of the LRA to be cancelled
+     * @return a {@link Response} object containing the optional error message 
+     */
+    @GET
+    @Path(PHASE_2)
+    public Response phase2(@HeaderParam(LRA_HEADER) URI lraId) {
+        lraTestService.getLRAClient().cancelLRA(lraId);
+
+        // assert compensate has been called
+        int compensations = lraMetricService.getMetric(LRAMetricType.Compensated, lraId, RecoveryResource.class.getName());
```

Comment:
I have a similar comment to Rudy, how do you know the implementation has finished cancelling the LRA. Perhaps a call to short delay is required here.

Reply:
ok